custom_random_forest.py

```diff
@@ -0,0 +1,227 @@
+        for tree_num in estimators:
+            with open(f'{self.tmp_dir}/{name}_{tree_num}_dump.pkl', 'rb') as md:
+                match name:
+                    case 'model':
+                        model, model_feature_idx = pickle.load(md)
+                        self.trees.append(model)
+                        self.feat_ids_by_tree.append(model_feature_idx)
+                    case 'proba':
+                        self.probas_y.append(pickle.load(md))
+
+    def fit(self, X, y, n_jobs):
+        self.classes_ = sorted(np.unique(y))
+
+        processes = [multiprocessing.Process(
+            target=self.fit_dts, args=(X, y, estimators)
+            ) for estimators in self.split_estimators(n_jobs)]
+        for proc in processes:
+            proc.start()
+        for proc in processes:
+            proc.join()
+
+        threads = [threading.Thread(
+            target=self.assemble_data_from_process, args=(estimators, 'model')
+            ) for estimators in self.split_estimators(n_jobs)]
+        for thread in threads:
+            thread.start()
+        for thread in threads:
+            thread.join()
+
+        return self
+
+    def predict_proba_dts(self, X, estimators):
+
+        for tree_num in estimators:
+            tree = self.trees[tree_num]
+            tree_proba = tree.predict_proba(X[:, self.feat_ids_by_tree[tree_num]])
+
+            with open(f'{self.tmp_dir}/proba_{tree_num}_dump.pkl', 'wb') as md:
+                pickle.dump(tree_proba, md)
+
+    def predict_proba(self, X, n_jobs):
+        self.probas_y.clear()
+        
+        threads = [threading.Thread(
+            target=self.predict_proba_dts, args=(X, estimators)
+            ) for estimators in self.split_estimators(n_jobs)]
+        for thread in threads:
+            thread.start()
+        for thread in threads:
+            thread.join()
+
+        threads = [threading.Thread(
+            target=self.assemble_data_from_process, args=(estimators, 'proba')
+            ) for estimators in self.split_estimators(n_jobs)]
+        for thread in threads:
+            thread.start()
+        for thread in threads:
+            thread.join()
+
+        return np.array(self.probas_y).mean(axis=0)
+
+    def predict(self, X, n_jobs):
+        probas = self.predict_proba(X, n_jobs)
+        predictions = np.argmax(probas, axis=1)
+        self.probas_y.clear()
+
+        return predictions
+
+
+class RandomForestClassifierCustomST(BaseEstimator):
+    def __init__(
+        self, n_estimators: int = 10, max_depth=None, max_features=None, random_state=SEED
+    ):
+        self.n_estimators = n_estimators
+        self.max_depth = max_depth
+        self.max_features = max_features
+        self.random_state = random_state
+
+        self.trees = []
+        self.feat_ids_by_tree = []
+
+    def fit_tree(self, X, y, estimator):
+        n_samples, n_features = X.shape
+
+        np.random.seed(seed=self.random_state + estimator)

```

Comment:
```suggestion
    def fit_tree(self, X, y, estimator_id)
        n_samples, n_features = X.shape

        np.random.seed(seed=self.random_state + estimator_id)
```
Если  я не путаю, там же именно номер лежит, а не само дерево?